Sources/GRPCInterceptors/Tracing/ClientOTelTracingInterceptor.swift

```diff
@@ -0,0 +1,196 @@
+          span.addEvent("Request ended")
+        }
+      }
+
+      var response = try await next(request, context)
+      switch response.accepted {
+      case .success(var success):
+        span.addEvent("Received response start")
+        span.attributes.rpc.grpcStatusCode = 0
+        if self.emitEventOnEachWrite {
+          let messageReceivedCounter = Atomic(1)
+          let onEachPartRecordingSequence = success.bodyParts.map { element in
+            var event = SpanEvent(name: "rpc.message")
+            event.attributes.rpc.messageType = "RECEIVED"
+            event.attributes.rpc.messageID =
+              messageReceivedCounter
+              .wrappingAdd(1, ordering: .sequentiallyConsistent)
+              .oldValue
+            span.addEvent(event)
+            return element
+          }
+
+          let onFinishRecordingSequence = OnFinishAsyncSequence(
+            wrapping: onEachPartRecordingSequence
+          ) {
+            span.addEvent("Received response end")
+          }
+
+          success.bodyParts = RPCAsyncSequence(wrapping: onFinishRecordingSequence)
+          response.accepted = .success(success)
+        } else {
+          let onFinishRecordingSequence = OnFinishAsyncSequence(wrapping: success.bodyParts) {
+            span.addEvent("Received response end")
+          }
+
+          success.bodyParts = RPCAsyncSequence(wrapping: onFinishRecordingSequence)
+          response.accepted = .success(success)
+        }
+
+      case .failure(let error):
+        span.attributes.rpc.grpcStatusCode = error.code.rawValue
+        span.setStatus(SpanStatus(code: .error))
+        span.addEvent("Received error response")
+        span.recordError(error)
+      }
+
+      return response
+    }
+  }
+
+  private func setOTelSpanAttributes(into span: any Span, context: ClientContext) {
+    span.attributes.rpc.system = "grpc"
+    span.attributes.rpc.service = context.descriptor.service.fullyQualifiedService
+    span.attributes.rpc.method = context.descriptor.method
+    span.attributes.rpc.serverAddress = self.serverHostname
+    span.attributes.rpc.networkTransport = self.networkTransportMethod
+
+    let peer = context.remotePeer
+    // We expect this address to be of either of these two formats:
+    // - <type>:<host>:<port> for ipv4 and ipv6 addresses
+    // - unix:<uds-pathname> for UNIX domain sockets
+    let components = peer.split(separator: ":")
+    if components.count == 2 {
+      // This is the UDS case
+      span.attributes.rpc.networkType = String(components[0])
+      span.attributes.rpc.networkPeerAddress = String(components[1])
+    } else if components.count == 3 {
+      // This is the ipv4 or ipv6 case
+      span.attributes.rpc.networkType = String(components[0])
+      span.attributes.rpc.networkPeerAddress = String(components[1])
+      span.attributes.rpc.networkPeerPort = Int(components[2])
+      span.attributes.rpc.serverPort = Int(components[2])
+    }
```

Comment:
Can we extract this logic into a helper to make it easier to test? I.e. something like:

```swift
struct AddressInfo {
  var networkType: String
  // etc.

  init(parsing peerInfo: String) {
    // ...
  }
}
```

It can then be re-used for the server interceptor.

Reply:
I've actually refactored this, but it ended up on the server PR https://github.com/grpc/grpc-swift-extras/pull/27
I can move it into this PR.